File: 872.叶子相似的树/872.叶子相似的树.cpp
/*
 * @Descroption: LeetCode 872. 叶子相似的树
 * @Author: EmoryHuang
 * @Date: 2021-04-25 15:01:36
 * @解题思路:
 * 通过先序遍历将叶子结点加入数组，而后比较两个数组的叶子节点是否相同
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
   public:
    void preOrder(TreeNode* root, vector<int>& leaf) {
        if (root == nullptr) return;
        if (!root->left && !root->right) {
            leaf.push_back(root->val);
            return;
        }
        preOrder(root->left, leaf);
        preOrder(root->right, leaf);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> leaf1, leaf2;
        preOrder(root1, leaf1);
        preOrder(root2, leaf2);
        if (leaf1.size() != leaf2.size()) return false;
        for (int i = 0; i < leaf1.size(); i++)
            if (leaf1[i] != leaf2[i]) return false;
        return true;
    }
};

File: Solution/872.叶子相似的树/872.叶子相似的树.java
/*
 * @Descroption: LeetCode 872. 叶子相似的树
 * @Author: EmoryHuang
 * @Date: 2023-07-15 10:35:15
 * @Method:
 * dfs分别记录两棵树的叶子节点
 */

import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> ls1 = new ArrayList<>();
        if (root1 != null) {
            dfs(root1, ls1);
        }

        List<Integer> ls2 = new ArrayList<>();
        if (root2 != null) {
            dfs(root2, ls2);
        }
        return ls1.equals(ls2);
    }

    void dfs(TreeNode root, List<Integer> ls) {
        if (root.left == null && root.right == null) {
            ls.add(root.val);
        } else {
            if (root.left != null) {
                dfs(root.left, ls);
            }
            if (root.right != null) {
                dfs(root.right, ls);
            }
        }
    }
}